### ISSUE TITLE
TLHC platform views on Android 14/15 stop updating on resume

### ISSUE THREAD

**Issue by evil159** [flutter/flutter#146499]
### Steps to reproduce

I've created a new project with flutter create and integrated a simple test view on Android side as per https://docs.flutter.dev/platform-integration/android/platform-views. 
Set up `FlutterEngine` cache as per Step 3 of this guide https://docs.flutter.dev/add-to-app/android/add-flutter-screen#step-3-optional-use-a-cached-flutterengine.

##### Steps to reproduce

1. Launch the app
1. Switch to Home Screen
1. Open an app or two(camera, phone app etc.)
    * This step can be skipped if `Developer option > Don't keep activities` is enabled
1. Go back to the sample app.

##### Environment

* Pixel 8 running AP31.240223.016.A3(Android 15 Preview)

Sample project: [platform_view_test 3.zip](https://github.com/flutter/flutter/files/14917706/platform_view_test.3.zip)

### Expected results

The platform view hosted by Flutter app is displayed and interactive.

### Actual results

* On Galaxy the platform view is not displayed, instead the place it used to occupy is rendered solid white.
* On Pixel the platform view is displayed, but not interacting(doesn't respond to touch).

### Code sample

<details open><summary>Code sample</summary>

Dart:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
      final Map<String, dynamic> creationParams = <String, dynamic>{};

    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text("App"),
      ),
      body: AndroidView(
        viewType: "nativeView",
        layoutDirection: TextDirection.ltr,
        creationParams: creationParams,
        creationParamsCodec: const StandardMessageCodec(),
      ),
    );
  }
}
```

Platform side:

```kotlin
class MyApplication : Application() {
    private lateinit var flutterEngine : FlutterEngine

    override fun onCreate() {
        super.onCreate()

        // Instantiate a FlutterEngine.
        flutterEngine = FlutterEngine(this)

        // Start executing Dart code to pre-warm the FlutterEngine.
        flutterEngine.dartExecutor.executeDartEntrypoint(
            DartExecutor.DartEntrypoint.createDefault()
        )

        // Cache the FlutterEngine to be used by FlutterActivity.
        FlutterEngineCache
            .getInstance()
            .put("my_engine_id", flutterEngine)
    }
}

class RootActivity: Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        startActivity(
            FlutterActivity.CachedEngineIntentBuilder(MainActivity::class.java, "my_engine_id")
                .build(this)
        )
        finish()
    }
}

class MainActivity: FlutterActivity() {
    override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
        super.configureFlutterEngine(flutterEngine)
        flutterEngine
            .platformViewsController
            .registry
            .registerViewFactory("nativeView", NativeViewFactory())
    }
}
```
</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/1478430/63054699-c025-43ca-a205-1ecc82eb2c99

</details>

### Logs

<details open><summary>Logs</summary>

```console
---------------------------- PROCESS STARTED (19824) for package com.example.platform_view_test ----------------------------
2024-04-09 14:01:09.947 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 171979766; UID 10274; state: ENABLED
2024-04-09 14:01:09.947 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 242716250; UID 10274; state: DISABLED
2024-04-09 14:01:09.959 19824-19824 ApplicationLoaders      com.example.platform_view_test       D  Returning zygote-cached class loader: /system_ext/framework/androidx.window.extensions.jar
2024-04-09 14:01:09.959 19824-19824 ApplicationLoaders      com.example.platform_view_test       D  Returning zygote-cached class loader: /system_ext/framework/androidx.window.sidecar.jar
2024-04-09 14:01:09.962 19824-19824 ziparchive              com.example.platform_view_test       W  Unable to open '/data/data/com.example.platform_view_test/code_cache/.overlay/base.apk/classes.dm': No such file or directory
2024-04-09 14:01:09.975 19824-19824 ziparchive              com.example.platform_view_test       W  Unable to open '/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/base.dm': No such file or directory
2024-04-09 14:01:09.975 19824-19824 ziparchive              com.example.platform_view_test       W  Unable to open '/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/base.dm': No such file or directory
2024-04-09 14:01:10.011 19824-19824 nativeloader            com.example.platform_view_test       D  Configuring clns-6 for other apk /data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/base.apk. target_sdk_version=33, uses_libraries=, library_path=/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/lib/arm64:/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/base.apk!/lib/arm64-v8a, permitted_path=/data:/mnt/expand:/data/user/0/com.example.platform_view_test
2024-04-09 14:01:10.024 19824-19824 GraphicsEnvironment     com.example.platform_view_test       V  Currently set values for:
2024-04-09 14:01:10.024 19824-19824 GraphicsEnvironment     com.example.platform_view_test       V    angle_gl_driver_selection_pkgs=[com.android.angle, com.linecorp.b612.android, com.campmobile.snow, com.google.android.apps.tachyon]
2024-04-09 14:01:10.024 19824-19824 GraphicsEnvironment     com.example.platform_view_test       V    angle_gl_driver_selection_values=[angle, native, native, native]
2024-04-09 14:01:10.024 19824-19824 GraphicsEnvironment     com.example.platform_view_test       V  com.example.platform_view_test is not listed in per-application setting
2024-04-09 14:01:10.025 19824-19824 GraphicsEnvironment     com.example.platform_view_test       V  Neither updatable production driver nor prerelease driver is supported.
2024-04-09 14:01:10.049 19824-19897 ResourceExtractor       com.example.platform_view_test       I  Found extracted resources res_timestamp-1-1712655907628
2024-04-09 14:01:10.052 19824-19896 nativeloader            com.example.platform_view_test       D  Load /data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/lib/arm64/libflutter.so using ns clns-6 from class loader (caller=<unknown>): ok
2024-04-09 14:01:10.059 19824-19824 tform_view_test         com.example.platform_view_test       W  type=1400 audit(0.0:18469): avc:  denied  { read } for  name="max_map_count" dev="proc" ino=6767069 scontext=u:r:untrusted_app_32:s0:c18,c257,c512,c768 tcontext=u:object_r:proc_max_map_count:s0 tclass=file permissive=0 app=com.example.platform_view_test
2024-04-09 14:01:10.133 19824-20001 vulkan                  com.example.platform_view_test       D  searching for layers in '/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/lib/arm64'
2024-04-09 14:01:10.142 19824-19941 flutter                 com.example.platform_view_test       I  The Dart VM service is listening on http://127.0.0.1:40127/VYwrTLx62Uk=/
2024-04-09 14:01:10.160 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 279646685; UID 10274; state: DISABLED
2024-04-09 14:01:10.191 19824-20001 vulkan                  com.example.platform_view_test       D  added global layer 'VK_LAYER_KHRONOS_validation' from library '/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/lib/arm64/libVkLayer_khronos_validation.so'
2024-04-09 14:01:10.216 19824-20001 vulkan                  com.example.platform_view_test       D  searching for layers in '/data/app/~~Jwigw4BUI7MoNTTzWd31uw==/com.example.platform_view_test-mcIrPFuHK3VSxqnVEsDqsg==/base.apk!/lib/arm64-v8a'
2024-04-09 14:01:10.228 19824-19824 tform_view_test         com.example.platform_view_test       W  Accessing hidden method Landroid/view/accessibility/AccessibilityNodeInfo;->getSourceNodeId()J (unsupported,test-api, reflection, allowed)
2024-04-09 14:01:10.228 19824-19824 tform_view_test         com.example.platform_view_test       W  Accessing hidden method Landroid/view/accessibility/AccessibilityRecord;->getSourceNodeId()J (unsupported, reflection, allowed)
2024-04-09 14:01:10.228 19824-19824 tform_view_test         com.example.platform_view_test       W  Accessing hidden field Landroid/view/accessibility/AccessibilityNodeInfo;->mChildNodeIds:Landroid/util/LongArray; (unsupported, reflection, allowed)
2024-04-09 14:01:10.228 19824-19824 tform_view_test         com.example.platform_view_test       W  Accessing hidden method Landroid/util/LongArray;->get(I)J (unsupported, reflection, allowed)
2024-04-09 14:01:10.247 19824-19824 ContentCaptureHelper    com.example.platform_view_test       I  Setting logging level to OFF
2024-04-09 14:01:10.254 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 237531167; UID 10274; state: DISABLED
2024-04-09 14:01:10.301 19824-19824 WindowExtensionsImpl    com.example.platform_view_test       I  Initializing Window Extensions.
2024-04-09 14:01:10.581 19824-19855 tform_view_test         com.example.platform_view_test       I  Compiler allocated 4683KB to compile void android.view.ViewRootImpl.performTraversals()
2024-04-09 14:01:10.664 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 171228096; UID 10274; state: ENABLED
2024-04-09 14:01:10.666 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 210923482; UID 10274; state: ENABLED
2024-04-09 14:01:10.666 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 63938206; UID 10274; state: DISABLED
2024-04-09 14:01:10.703 19824-19855 tform_view_test         com.example.platform_view_test       I  Compiler allocated 4213KB to compile void android.widget.TextView.<init>(android.content.Context, android.util.AttributeSet, int, int)
2024-04-09 14:01:10.705 19824-19824 PlatformViewsController com.example.platform_view_test       I  Hosting view in view hierarchy for platform view: 0
2024-04-09 14:01:10.706 19824-19824 PlatformViewsController com.example.platform_view_test       I  PlatformView is using ImageReader backend
2024-04-09 14:01:10.706 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 236825255; UID 10274; state: DISABLED
2024-04-09 14:01:10.770 19824-19824 ImageTextu...istryEntry com.example.platform_view_test       E  Dropping PlatformView Frame
2024-04-09 14:01:10.777 19824-19824 ContentCaptureHelper    com.example.platform_view_test       I  Setting logging level to OFF
2024-04-09 14:01:11.933 19824-19824 ImageTextu...istryEntry com.example.platform_view_test       E  Dropping PlatformView Frame
2024-04-09 14:01:14.605 19824-19824 VRI[MainActivity]       com.example.platform_view_test       D  visibilityChanged oldVisibility=true newVisibility=false
2024-04-09 14:01:14.677 19824-19824 WindowOnBackDispatcher  com.example.platform_view_test       W  sendCancelIfRunning: isInProgress=falsecallback=android.view.ViewRootImpl$$ExternalSyntheticLambda11@e05917c
2024-04-09 14:01:17.388 19824-19824 ContentCaptureHelper    com.example.platform_view_test       I  Setting logging level to OFF
2024-04-09 14:01:18.676 19824-19824 Compatibil...geReporter com.example.platform_view_test       D  Compat change id reported: 289878283; UID 10274; state: DISABLED
---------------------------- PROCESS ENDED (19824) for package com.example.platform_view_test ----------------------------
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.5, on macOS 14.4.1 23E224 darwin-arm64, locale en-FI)
    • Flutter version 3.19.5 on channel stable at /Users/romanlaitarenko/Downloads/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 300451adae (12 days ago), 2024-03-27 21:54:07 -0500
    • Engine revision e76c956498
    • Dart version 3.3.3
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/romanlaitarenko//Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/romanlaitarenko//Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] Connected device (6 available)
    • Pixel 8 (mobile)            • 37280DLJH000VR                       • android-arm64  • Android 14 (API 34)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        • android-arm64  • Android 14 (API 34) (emulator)
    • Roman’s iPhone (2) (mobile) • 00008110-00110D1614F9801E            • ios            • iOS 17.4.1 21E236
    • iPhone 15 Pro (mobile)      • AB1C69B3-AE98-458C-BD91-BEFAFF2357B0 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-17-4 (simulator)
    • macOS (desktop)             • macos                                • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                • chrome                               • web-javascript • Google Chrome 123.0.6312.59

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @evil159 
Most likely you are hitting https://github.com/flutter/flutter/issues/139630 that you can check and confirm since you are also seeing similar behavior you mentioned on Samsung device running Android 14.

https://github.com/flutter/flutter/issues/139630#issuecomment-2007673827 for your reference.

**Comment by evil159** [comment]
@darshankawar Indeed the issue on Samsung can be explained by that, but then there is a separate bug on Pixels.

**Comment by evil159** [comment]
@darshankawar I've updated this issue accordingly, could you take a look at this again?

I don't believe that the fix that Samsung is preparing for Galaxy will fix the issue on Pixel phones.

**Comment by johnmccutchan** [comment]
I've been able to reproduce this issue on a Pixel 7 Pro.

**Comment by johnmccutchan** [comment]
I added a lot of traces to the engine and this is my current understanding of the bug:

After we resume the ImageReader.OnImageAvailableListener callback never fires again.

We continue to render frames and notify the engine that a frame is available but the image reader is never notified and when we ask the imagereader for its latest frame null is returned.

**Comment by johnmccutchan** [comment]
Another note:

The difference between stable and master is because https://github.com/flutter/engine/pull/50792 removed the cached first frame. That is,

When stable resumes we use the last frame rendered before we were backgrounded.
When main resumes we have nothing to render and so you see the Flutter app's background color.

Both stable and master would be fixed if `ImageReader.OnImageAvailableListener` was firing as expected.

**Comment by reidbaker** [comment]
Was not able to reproduce with a pixel fold running an unreleased version of android 14. 

pipit-userdebug UpsideDownCake MASTER 93172441 dev-keys

**Comment by reidbaker** [comment]
Was not able to reproduce with Nexus 6 running android 7.1.2 build number n2g47w

**Comment by reidbaker** [comment]
I was able to reproduce on a pixel 6 running android 14 build number AP1A.240405.002

**Comment by johnmccutchan** [comment]
Today's update:

We've tried versions of Flutter back to 2022 and none of them work correctly on Android 14 (to be more precise: there are some Android 14 builds that work and some that do not, early Android 15 builds are broken too).

We have not reproduced this issue on Android < 14.

Simplest reproduction steps:

Open sample app
Background sample app.
Launch the camera.
Background camera.
Use task switcher to select sample app.
We are trying to bisect on the Android side to understand more.

**Comment by johnmccutchan** [comment]
Also, it does not appear related to having a cached engine.

**Comment by johnmccutchan** [comment]
Also, rotating the phone will fix the app.

**Comment by reidbaker** [comment]
A coworker bisected this issue using an emulator and was able to find that the problematic android change happened between between build 10480234 (from 7/11/2023) and build 10488473 (from 7/12/2023). 

Links for googlers 
10480234: 
* Bug for the build https://b.corp.google.com/issues/290863690
* Build info https://coastguard.corp.google.com/build/ZP1A.230712.001/
* Bugs fixed https://groups.google.com/a/google.com/g/android-build-police/c/ijdTOnVtluI/m/Dr3UlwCWCAAJ 

For the second build I can't find any records. 
https://groups.google.com/a/google.com/g/android-build-police/search?q=after%3A2023-07-11%20before%3A2023-07-13%20

**Comment by reidbaker** [comment]
Build names appear to be 
10480234  -> ZP1A.230712.001
10488473 -> ZP1A.230713.001
https://android-build.corp.google.com/build_explorer/branch/git_trunk-release/?gridSize=20&selectionType=START_BUILD_WINDOW&startBuildId=10488473&numBuilds=20&buildType=default

I think this url will let us search for changes but there were 1000 that landed. 
https://android-build.corp.google.com/build_explorer/branch_dashboard/git_trunk-release/build_id/10488473/?gridSize=20&activeTarget=akita-user

**Comment by reidbaker** [comment]
From the android team a better way to search commits. 
https://android-build.corp.google.com/range_search/cls/from_id/10488473/to_id/10480234/?s=menu&includeTo=0&includeFrom=1

**Comment by reidbaker** [comment]
Current belief is that this is an android regression that appeared during android 14 and we have escalated to the android team. Internal bug b/330184547

**Comment by reidbaker** [comment]
https://googleplex-android.googlesource.com/platform%2Fframeworks%2Fbase/+/20a4d68338ca000e3ee0c5c71a01552eab1061e1
the commit we think is the source of the error.

**Comment by evil159** [comment]
Is there any workaround for this issue? Any way to make that `ImageReader.OnImageAvailableListener` firing again?

**Comment by reidbaker** [comment]
We don't know of any work around that apps can trigger as of this moment if we do we will add it to this bug.  The android team has informed us of a way flutter can avoid this error and we are working on implementing a fix. That fix will require apps to adopt a new version of flutter.

**Comment by reidbaker** [comment]
The android team has confirmed that the commit we thought was the issue was indeed the problem and has a fix on their side. Unfortunately android source changes take a long time to ship and live in the wild for a longer time. We are working to avoid the classes that are bugged and will update this issue when there is something app developers can do.

**Comment by yh-luo** [comment]
WebView is not functioning properly due to this issue. For my app, multiple Samsung and Pixel users have reported that the app displays a white screen upon resume when the UI utilizes WebView, as noted in #139039.

> Also, rotating the phone will fix the app.

I tested this on a Pixel 7 running Android 14, but unfortunately, it did not resolve the issue :cry:

**Comment by phcd30** [comment]
This is crazy that flutter webview does not work on the latest version of Android. The response in the [previous thread](https://github.com/flutter/flutter/issues/139039) was incredibly poor as well, very dismissive. 

"Locked as spam" on an issue this critical. I'm sure a lot of other people using or looking at Flutter will be steered away by this type of developer interaction. 

We've just launched a new app on flutter and every user with a recent Samsung or Android it seems are getting a lot of white screens, their only option is to close and open the app again. I'd put this in the P-Minus-Zero category of priorities if Flutter is to be taken seriously. 

![image](https://github.com/flutter/flutter/assets/1768859/97d1d887-d571-46c6-bc9a-14cd4a84ac99)

**Comment by reidbaker** [comment]
@phcd30 there is a pull request open linked on this pr with a fix for flutter. https://github.com/flutter/engine/pull/52370

**Comment by phcd30** [comment]
Do you think it will be fixed with a flutter update without the need to wait for Android 15 or a patch to Android 14?

The other thread starts with this statement - "Android 15 will include the fix. Unclear if Android 14 will be patched."

**Comment by Hixie** [comment]
@phcd30 Flutter is an open source project, many of us volunteer to work on this project in our free time; others are employed by companies that donate their resources and time to the project. We all have our own list of what is most critical to us and what we need to work on next to address our needs.

If this is "P-Minus-Zero" for you, your best option is to contribute a fix or hire someone to work on a fix. We would certainly welcome you or someone you hire into the community and would love to help you become productive contributors in whatever way we can. You can find more in [our contributing guide](https://github.com/flutter/flutter/blob/main/CONTRIBUTING.md).

I encourage you to reread your comment and think how it might be received by someone whom you are not paying but who has nonetheless contributed their time to create a product that you now rely on.

> Also, if your logging and alerting are on point you should be lighting up like a Christmas tree, not asking people to repro on different devices.

For privacy reasons, we do not log anything from applications shipped with Flutter, so we have to rely on folks who do log information from their users to report that information to us.

**Comment by phcd30** [comment]
Happy to test this fix if possible? 

My feedback is constructive, take it as it's intended. If Flutter is viewing it's as a hobby project, as opposed to the marketing on the flutter home page - 

![image](https://github.com/flutter/flutter/assets/1768859/6ba206a6-8b29-4931-9f02-16a5dd8565f2)


 it would be irresponsible for anyone to consider it. Please update the marketing to reflect the views and responses on this repo.

**Comment by Hixie** [comment]
Please review our [code of conduct](https://github.com/flutter/flutter/blob/main/CODE_OF_CONDUCT.md), especially the part relating to respect people's work and being kind and welcoming. I have blocked your account for seven days.

**Comment by Hixie** [comment]
(Please move discussion about how Flutter is marketed to another issue or to a Discord thread; it is off topic for this issue. I'm happy to discuss that particular topic, but we should keep issues focused on the topic at hand, as per <https://github.com/flutter/flutter/wiki/Issue-hygiene#issues-are-not-always-the-best-venue-for-discussions>. If you do wish to have that discussion, please cc me! Thanks.)

**Comment by neilgrattan** [comment]
"Please consult us about how we deceived you and how we can improve this going forward" is not an appropriate response.

**Comment by reidbaker** [comment]
We know this issue is serious and have spent multiple engineering weeks tracking down the root cause. 
Current recap for those that are following. 

On Android 14 platform views encounter an Android bug where we stop getting draw information from android when memory trim is called, usually the app is in the background. This causes platform views to render transparent BUT the actual underlying views are still there and are interactable. This set of trigger conditions is not unique to flutter. The apis we use would also fail for Android apps. Additionally there was a change between flutter 3.19 and flutter master (probably 3.20) where for platform views we no longer cache the last frame and draw it onResume until we get the surface information from android. Before that previous frame helped hide this issue since the underlying android views were still interactable. 

To the best of my knowledge, this issue does not appear when using the app regularly. It does not appear when backgrounded unless under memory pressure. Rotating the phone clears the condition. Evicting the app and relaunching clears the condition. 

In order to detect the bug with an automated test the test would have to 1) Launch an app with a platform view, 2) background the app, 3) trigger memory trim, 4) relaunch the app 5) Scroll or some other action to move items 6) Take a screenshot and pixel compare. 
If you do not do step 5 then the test would continue to work because the last frame would have been preserved and you could still click the button and check for the callback or any other touch interaction. If you did step 5 but did not pixel compare a screenshot then the test would also have passed again because the underlying buttons are there but not visible/updating. As a general rule combining screenshots with scrolling is extremely flakey so we have not included those types of tests. 

Also extending this bug is that it was reported as a samsung specific issue and the contributors working on this issue did not have access to samsung code to see how it differed. Samsung confirmed reproducibility but cross company interactions are significantly slower and then we were told Samsung had a fix but the code was not something we could review. I will admit my own fault that I assumed that Pixel phones would always have the newest version of android so I did not consider that this could have been a pure Android issue. 

When we eventually tried to prove or disprove that Android had the bug we bisected a years worth of google internal android builds and narrowed it down to 2 days worth of commits (6 thousand commits) then searched for commits that could have touched the code at issue. The android commit at issue is linked above. 

If you are here you probably care most about the fix. 
Android has a fix committed to their upstream but their release process is slower than we know our customers want. We have confirmed that Android's fix will be in Android 15. We are escalating to see if we(flutter) can get them(Android) to backport the fix to future patch versions of Android 14, if that is approved and I am allowed to share android release information I expect to update this bug . @johnmccutchan has a pull [request up now](https://github.com/flutter/engine/pull/52370) that has a work around to avoid the bug. Once that pr is merged, flutter plans to backport that pr into older flutter engine builds. This bug will be updated with which flutter versions have the fix when they land and the version numbers of those builds. Finally if you are running a custom engine or an engine older than we backport you can apply the patch yourself and build with a custom engine.

**Comment by reidbaker** [comment]
A way to force the background trim condition `adb shell am send-trim-memory <YOUR PACKAGE NAME>  BACKGROUND

**Comment by amrgetment** [comment]
That’s a great effort from your team, people should see how the team is squeezing their brains to solve the flutter issues and make it a better framework
Thanks a lot

**Comment by OnClickListener2048** [comment]
This is a great job and thanks all for making this progress so far.

**Comment by lukehutch** [comment]
> This causes platform views to render transparent BUT the actual underlying views are still there and are interactable.
> 
> To the best of my knowledge, this issue does not appear when using the app regularly. It does not appear when backgrounded unless under memory pressure. Rotating the phone clears the condition. Evicting the app and relaunching clears the condition.

I see this problem on both Pixel 4 and Pixel 8 Pro, both on Android 14. However, my symptoms are somewhat different than this description.

1. I get a "black screen of death" when backgrounding the app for some period of time, e.g. 5 minutes (presumably under memory pressure, but not always in any obvious way), then switching back to the app.
2. The app is not responsive in any way -- I sometimes (but not always) get an ANR dialog if I leave the app in the "black screen of death" view for a few seconds.
3. I cannot connect a debugger to the running Flutter instance, which confirms that the app is not responsive under the black screen.
4. Rotating the screen does not clear the condition in my case.
5. The Back button pops the whole app, rather than what would be the current view in the app, which is another indication that the app is unresponsive.

**Comment by alpha2048** [comment]
Thank you to Flutter team for their hard work.
But when will the work around be available on the stable channel? Our users using Pixel are still having trouble.

**Comment by reidbaker** [comment]
3.22 has had the fix pr cherry picked into the engine. We are working on creating a new framework build that has the updated engine code. 3.22 is scheduled to land as stable sometime around the Google IO timeframe (next week). 

Standard response below. 

This issue has been fixed by https://github.com/flutter/engine/pull/52370  To determine what versions of Flutter contain the fix, please see [where’s my commit](https://github.com/flutter/flutter/wiki/Where's-my-Commit%3F). If the commit has no release tag, it is currently only available in the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#flutters-channel).

**Comment by evil159** [comment]
Is this fixed in 3.22.0? Cannot find anything about https://github.com/flutter/engine/pull/52370 in the [release notes](https://docs.flutter.dev/release/release-notes/release-notes-3.22.0).

**Comment by yh-luo** [comment]
I think the fix is merged in [3793869](https://github.com/flutter/engine/commit/379386938a92e332abf9ad8cb21d67308c11fefe).

I found that under [flutter/engine/3.22.0](https://github.com/flutter/engine/commits/3.22.0/). Also the fix was mentioned in [What’s new in Flutter 3.22: Platform views improvements](https://medium.com/flutter/whats-new-in-flutter-3-22-fbde6c164fe3).

**Comment by mr-pant** [comment]
I see that this issue is fixed with Flutter 3.22.0 for the mentioned steps, but I see this happening following different steps instead. Now if the platform view is your initial page and you press **back button** to move app to background, on launching again, you will see that the blank issue is reproducible. So looks like issue is still not fully fixed.

Another interesting thing is that after this, if you follow the steps for old issue i.e - switch from another recent app to sample app from recent apps launcher, the platform view starts working again.

**Comment by reidbaker** [comment]
Promised update on the android fix. 
The change to fix existing flutter apps (pre 3.22) on android 14 has been cherry picked into the July release of Pixel. I am still working on the escalation for other manufacturers.

**Comment by reidbaker** [comment]
Promised update on Android fix for non google manufacturers. The patch could not cherry pick cleanly into AOSP and thus has been deemed too risky to introduce as an option to other OEMs. We have been told that Android 15 will have the fix. 

The first patch release for 3.22 should be out wednesday 2024-06-03 which will contain the last known flutter change to avoid the underlying android issue.

**Comment by amrgetment** [comment]
So, Samsung devices with Android 14 will crash or Flutter version 3.22.2 will fix it?

**Comment by reidbaker** [comment]
I am actively working on a postmortem and hoping it gets approval to be public. 

@amrgetment yes, 3.22.0 will fix most cases. There was a corner case for multi activity apps where one app was not a flutter activity that is fixed in 3.22.2. I will note that apps dont crash, they fail to get new pixels rendered for platform views.

**Comment by reidbaker** [comment]
> I am actively working on a postmortem and hoping it gets approval to be public.
> 
> @amrgetment yes, 3.22.0 will fix most cases. There was a corner case for multi activity apps where one app was not a flutter activity that is fixed in 3.22.2. I will note that apps dont crash, they fail to get new pixels rendered for platform views.

Post mortem published [docs/postmortems/Postmortem-Platform-View-android-14-regression](https://github.com/flutter/flutter/blob/master/docs/postmortems/Postmortem-Platform-View-android-14-regression)
